frontends/PyCDE/src/pycde/matrix.py

```diff
@@ -0,0 +1,264 @@
+    self = np.ndarray.__new__(cls, shape=shape, dtype=object)
+
+    if name is None:
+      name = "matrix"  # todo: require name?
+    self.name = name
+    self.pycde_dtype = dtype
+
+    # The SSA value of this matrix after it has been materialized.
+    # Once set, the matrix is immutable.
+    self.circt_output = None
+
+    if from_value is not None:
+      # PyCDE and numpy do not play nicely when doing np.arr(self.circt_to_arr(...))
+      # but individual assignments work.
+      target_shape = self.target_shape_for_idxs(0)
+      value_arr = self.circt_to_arr(from_value, target_shape)
+      for i, v in enumerate(value_arr):
+        super().__setitem__(self, i, v)
+
+    return self
+
+  @lru_cache
+  def get_constant(self, value: int, width: int = 32):
+    """ Get an IR constant backed by a constant cache."""
+    return hw.ConstantOp(ir.IntegerType.get_signless(width), value)
+
+  def circt_to_arr(self, value, target_shape):
+    """Converts a CIRCT value into a numpy array."""
+    if isinstance(value, BitVectorValue) and isinstance(target_shape.dtype,
+                                                        BitVectorType):
+      # Direct match on the target shape, which is not an array?
+      if value.type == target_shape.dtype and target_shape.num_dims == 0:
+        return value
+
+      # Is it feasible to extract values to the target shape?
+      if target_shape.num_dims > 1:
+        raise ValueError(
+            f"Cannot extract BitVectorValue of type {value.type} to a multi-dimensional array of type {target_shape.type}."
+        )
+
+      target_shape_bits = (target_shape.dims[0] if len(target_shape.dims) else
+                           1) * target_shape.dtype.width
+      if target_shape_bits != value.type.width:
+        raise ValueError(
+            f"Width mismatch between provided BitVectorValue ({value.type}) and target shape ({target_shape.type})."
+        )
+
+      # Extract to the target type
+      n = len(value) / target_shape.dtype.width
+      if n != int(n) or n < 1:
+        raise ValueError("Bitvector must be a multiple of the provided dtype")
+      n = int(n)
+      slice_elem_width = int(len(value) / n)
+      arr = []
+      if n == 1:
+        return value
+      else:
+        for i in range(n):
+          startbit = i * slice_elem_width
+          endbit = i * slice_elem_width + slice_elem_width
+          arr.append(value[startbit:endbit])
+    elif isinstance(value, ListValue):
+      # Recursively convert the list.
+      arr = []
+      for i in range(value.type.size):
+        # Pop the outer dimension of the target shape.
+        inner_dims = target_shape.dims.copy()[1:]
+        arr.append(
+            self.circt_to_arr(value[i],
+                              TargetShape(inner_dims, target_shape.dtype)))
+    else:
+      raise ValueError(f"Cannot convert value {value} to numpy array.")
+
+    return arr
+
+  def target_shape_for_idxs(self, idxs):
```

Comment:
Staticmethod?

Reply:
Uses `self.pycde_dtype` to reference the given instance's dtype.